### PR TITLE
Auto-reformat

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This PR contains two things. 

A shell script. If this shell script is used as a  hook to push, then all files will be reformatted before being pushed. If any change is not already commited, the push won't be accepted. This is not perfect, because it means you can't push anymore if you have some uncommitted change and you indeed wanted them to remain uncommitted. 
The script assume that you use android-studio, and that you can call it by simply using `android-studio`; I don't know how it would work on windows
The biggest problem is that this script does not work if android studio is opened. You can't ask from command line to reformat everything WITHOUT closing android studio first.

Lastly, since I didn't want actual reformatting to occur in this PR, I had to disable the hook in order to push this PR.

If we have access to InteliJ Idea or Android-studio on travis, we could do the same thing, ask to reformat, and if any change occured, reject.

The second and more interesting thing is that occurs is that it adds configuration for the plug-in "reformat on save". If you install this plug-in, each time you save a file, it'll be reformatted according to rules (assuming the file compiles), so you should never have to deal with reformatting again.

This hopefully would start to answer requests you had last time I wanted to apply a consistent formatting to files.
You can download this branch and run the hook to see how it would work and whether the formatting option please you or whether you'd want to change them.

I like having xml using more lines, this will lead to smaller conflict durring rebase.

### CODE REVIEW

File: tools/git-hooks/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+This directory contains hooks that contributors should call in their git repo.
```
Comment (david-allison-1) [inline tools/git-hooks/README.md:1]: Could you add a link/instructions to add this to git pre-commit hooks (I haven't used them before).

**Review by david-allison-1 — COMMENTED** [review]
Needs a second approval, this looks reasonable, but I don't have enough experience in the area to give a  review.

**Comment by mikehardy** [comment]
I've done hooks and stuff (checkout "husky" even https://github.com/typicode/husky so I'm not worried about that, just focused more on the back stack of PRs and release-2.13 then whitespace at the moment

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This current converts `public static` variables to `private static` if they're unused.

File: tools/git-hooks/README.md (hunk 1 above)
Comment (Arthur-Milchior) [reply tools/git-hooks/README.md:1]: Done

**Comment by Arthur-Milchior** [comment]
I'm really not looking forward to have it accepted immediately. It makes a huge amount of change.
I'm going over some changes in a lot of smaller PR so that we can consider each change individually, or ignore the one we don't want. I believe that this should wait until we've manually cleaned all warning that android studio find.

My goal here is that you can test this version and give your opinion about whether it's a good idea or not. Or even whether it seems something that can be enforced in travis (since it requires a version of IntelliJ to work, I'm not sure that it's okay)


> This current converts public static variables to private static if they're unused.

Most actions can be disabled. I'd suggest discussing the big picture, whether it seems a good way to approach the problem, before discussing exactly what fine tuning we want for IntelliJ's option

**Comment by david-allison-1** [comment]
I'm looking into this more, 

Personally: it'd be ideal to run the formatter on each hunk in a commit, so we get incremental improvements in the codebase and no longer need to review. 

Useful links:

https://www.jetbrains.com/help/idea/command-line-formatter.html

https://github.com/JetBrains/intellij-community/blob/master/platform/lang-impl/src/com/intellij/formatting/commandLine/FormatterStarter.java

**Comment by david-allison-1** [comment]
Eclipse's ADT allows formatting on a per-line basis, had a spike, avoiding for now as it's likely not going to be the route we take.

```
File file = path.toFile();
        // ToolFactory

        Map<String, String> options = new HashMap<>();

        options.put(DefaultCodeFormatterConstants.FORMATTER_COMMENT_FORMAT_BLOCK_COMMENT, "FALSE");
        options.put(DefaultCodeFormatterConstants.FORMATTER_COMMENT_FORMAT_JAVADOC_COMMENT, "FALSE");
        options.put(DefaultCodeFormatterConstants.FORMATTER_COMMENT_FORMAT_LINE_COMMENT, "FALSE");
        options.put(DefaultCodeFormatterConstants.FORMATTER_TAB_CHAR, JavaCore.SPACE);

        int mNoAlignment = DefaultCodeFormatterOptions.Alignment.M_NO_ALIGNMENT;
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_ADDITIVE_OPERATOR, String.valueOf(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_STRING_CONCATENATION, String.valueOf(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_ARGUMENTS_IN_ALLOCATION_EXPRESSION, String.valueOf(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_ARGUMENTS_IN_ANNOTATION, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_ARGUMENTS_IN_METHOD_INVOCATION, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_ARGUMENTS_IN_ENUM_CONSTANT, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_COMPACT_IF, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_TYPE_PARAMETERS, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_TYPE_ARGUMENTS, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_PARAMETERS_IN_METHOD_DECLARATION, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_EXPRESSIONS_IN_FOR_LOOP_HEADER, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_SHIFT_OPERATOR, Integer.toString(mNoAlignment));
        options.put(DefaultCodeFormatterConstants.FORMATTER_WRAP_BEFORE_STRING_CONCATENATION, "false");
        options.put(DefaultCodeFormatterConstants.FORMATTER_WRAP_OUTER_EXPRESSIONS_WHEN_NESTED, "false");

        // These are bad - means { will not add a new
        // options.put(DefaultCodeFormatterConstants.FORMATTER_INDENT_STATEMENTS_COMPARE_TO_BLOCK, "false");
        // options.put(DefaultCodeFormatterConstants.FORMATTER_INDENT_STATEMENTS_COMPARE_TO_BODY, "false");
        //ptions.put(DefaultCodeFormatterConstants.INDENT, "false");
        //options.put(DefaultCodeFormatterConstants.FORMATTER_INDENT_STATEMENTS_COMPARE_TO_BODY, "false");



/*
        indent_statements_compare_to_block = true
        indent_statements_compare_to_body = true*/

/*        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_, Integer.toString(M_NO_ALIGNMENT));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_, Integer.toString(M_NO_ALIGNMENT));
        options.put(DefaultCodeFormatterConstants.FORMATTER_ALIGNMENT_FOR_, Integer.toString(M_NO_ALIGNMENT));*/
/*        options.put(DefaultCodeFormatterConstants.FORMATTER_INSERT_SPACE_AFTER_STRING_CONCATENATION, "false");
        options.put(DefaultCodeFormatterConstants.FORMATTER_INSERT_SPACE_BEFORE_STRING_CONCATENATION, "false");*/



        // options.put(DefaultCodeFormatterConstants.FORMATTER_LINE_SPLIT, "300");


        DefaultCodeFormatter codeFormatter = new DefaultCodeFormatter();
        codeFormatter.setOptions(options);

        IDocument doc = new Document();
        try {
            // read the file
            String contents = new String(org.eclipse.jdt.internal.compiler.util.Util.getFileCharContent(file, null));
            // format the file (the meat and potatoes)
            doc.set(contents);
            int kind = (file.getName().equals(IModule.MODULE_INFO_JAVA)? CodeFormatter.K_MODULE_INFO
                    : CodeFormatter.K_COMPILATION_UNIT) | CodeFormatter.F_INCLUDE_COMMENTS;
            TextEdit edit = codeFormatter.format(kind, contents, 0, contents.length(), 0, null);
            if (edit != null) {
                edit.apply(doc);
            } else {
                return;
            }

            // write the file
            final BufferedWriter out = new BufferedWriter(new FileWriter(file));
            try {
                out.write(doc.get());
                out.flush();
            } finally {
                try {
                    out.close();
                } catch (IOException e) {
                    /* ignore */
                }
            }
        } catch (IOException e) {

        } catch (BadLocationException e) {

        }
```

**Pull request closed without merging**